crates/driver/src/domain/competition/solution/mod.rs

```diff
@@ -512,36 +515,56 @@ fn scaling_factor(first: &Prices, second: &Prices) -> Option<BigRational> {
     }
 }
 
-/// A unique reference to a specific solution. Sings IDs are generated by the
-/// solver and need to be unique within a single round of competition. If the
-/// driver merges solutions it combines individual IDs. This reference is only
-/// important in the communication between the driver and the solver, and it is
-/// not used by the protocol.
+/// A unique reference to a specific solution which consists of 2 parts:
+/// 1. Globally unique ID used to communicate with the protocol. Global
+///    uniquenes is enforced by the constructors.
+/// 2. List of merged sub ids. Each sub id was generated by the solver and only
+///    has to be unique within an auction run loop. If this list contains only a
+///    single id it means this Id belongs to an unmodified solution provided as
+///    is by the solver. If it contains multiple sub ids multiple base solutions
+///    have been merged into a bigger one.
 #[derive(Debug, Clone, Hash, PartialEq, Eq)]
-pub enum Id {
-    Single(u64),
-    Merged(Vec<u64>),
+pub struct Id {
+    id: u64,
+    merged_solutions: Vec<u64>,
 }
 
-impl From<u64> for Id {
-    fn from(value: u64) -> Self {
-        Id::Single(value)
+impl Id {
+    fn next_global_id() -> u64 {
+        static COUNTER: AtomicU64 = AtomicU64::new(0);
+        COUNTER.fetch_add(1, Ordering::Relaxed)
     }
-}
 
-impl Id {
-    /// Returns the number of solutions that has gone into merging this
-    /// solution.
-    pub fn count_merges(&self) -> usize {
-        self.ids().len()
+    pub fn new(solution: u64) -> Self {
+        Self {
+            id: Self::next_global_id(),
+            merged_solutions: vec![solution],
+        }
     }
 
-    pub fn ids(&self) -> Vec<u64> {
-        match self {
-            Id::Single(id) => vec![*id],
-            Id::Merged(ids) => ids.clone(),
+    pub fn new_merged(first: &Self, second: &Self) -> Self {
+        let merged_solutions = first
+            .solutions()
+            .iter()
+            .chain(second.solutions().iter())
+            .copied()
+            .collect();
+        Self {
+            id: Self::next_global_id(),
+            merged_solutions,
         }
     }
+
+    /// Globally unique id communicated to the protocol.
+    pub fn get(&self) -> u64 {
```

Comment:
Resetting this Id on restarts is fine since it's only used to let the autopilot exactly specify which solution the driver should execute. When the driver restarts and the number resets it will not have any solution in memory anyway so it should be fine.
Should we ever chose to persist solutions before restarting a driver we could consider changing this. Even if it only matters when a driver wins with solution `0`, restarts, wins again with solution `0` and then gets told to execute solution `0`. In that case the driver wouldn't know which solution `0` was intended.